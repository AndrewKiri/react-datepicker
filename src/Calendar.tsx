import * as React from 'react';
import * as moment from 'moment';
import CalendarHead from './CalendarHead';
import CalendarBody from './CalendarBody';
import 'moment/locale/ko';
import '../assets/styles/calendar.scss';


interface Props {
  headerFormat?: string
  selected?: moment.Moment[]
  startDay?: moment.Moment
  endDay?: moment.Moment
  locale?: string
  onChange?: (date: moment.Moment) => void
  customDayClass?: (date: moment.Moment) => string | string[]
  customDayText?: (date: moment.Moment) => string
}

interface State {
  current: moment.Moment
}

class Calendar extends React.Component<Props, State>{

  public static defaultProps = {
    headerFormat: "YYYY년 MM월",
    multiSelect: false
  }

  constructor(props: Props){
    super(props);
    this.state = {
      current: moment(),
    };
    moment.locale(props.locale);
  }

  handlePrev = () => {
    this.setState({
      current: this.state.current.subtract(1, 'months')
    });
  }

  handleNext = () => {
    this.setState({
      current: this.state.current.add(1, 'months')
    });
  }

  render() {
    const {
      headerFormat,
      customDayClass,
      customDayText,
      selected,
      onChange,
      startDay,
      endDay
    } = this.props;

    const {
      current,
    } = this.state;


    return (
      <div className="calendar">
        <CalendarHead
          onPrev={this.handlePrev} 
          onNext={this.handleNext} 
          title={current.format(headerFormat)}
        />
        <CalendarBody
          current={current} 
          selected={selected}
          startDay={startDay}
          endDay={endDay}
          onChange={onChange}
          customDayClass={customDayClass}
          customDayText={customDayText}
        />
      </div>
    )
  }
}

export default Calendar;
