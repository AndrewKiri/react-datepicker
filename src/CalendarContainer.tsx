import * as React from 'react';
import * as moment from 'moment';
import CalendarHead from './CalendarHead';
import CalendarBody from './CalendarBody';
import classNames from 'classnames';
import 'moment/locale/ko';

export interface InheritProps {
  headerFormat?: string
  locale?: string
  selected?: moment.Moment[]
  startDay?: moment.Moment
  endDay?: moment.Moment
  current: moment.Moment
  onChange?: (date: moment.Moment) => void
  customDayClass?: (date: moment.Moment) => string | string[]
  customDayText?: (date: moment.Moment) => string
  show?: boolean
  prevIcon?: boolean
  nextIcon?: boolean
}

interface PrivateProps {
  onPrev?: () => void
  onNext?: () => void
}

type Props = InheritProps & PrivateProps

class CalendarContainer extends React.Component<Props>{

  public static defaultProps = {
    headerFormat: "YYYY년 MM월",
    show: true,
    current: moment()
  }

  constructor(props: Props){
    super(props);
    moment.locale(props.locale);
  }


  render() {
    const {
      headerFormat,
      customDayClass,
      customDayText,
      selected,
      onChange,
      startDay,
      endDay,
      prevIcon,
      nextIcon,
      onPrev,
      onNext,
      show,
      current
    } = this.props;


    const calendarClass = classNames('calendar__container', {
      'calendar--show': show
    });

    return (
      <div className={calendarClass}>
        <CalendarHead
          onPrev={onPrev} 
          onNext={onNext} 
          prevIcon={prevIcon}
          nextIcon={nextIcon}
          title={current.format(headerFormat)}
        />
        <CalendarBody
          current={current}
          selected={selected}
          startDay={startDay}
          endDay={endDay}
          onChange={onChange}
          customDayClass={customDayClass}
          customDayText={customDayText}
        />
      </div>
    )
  }
}


export default CalendarContainer;
