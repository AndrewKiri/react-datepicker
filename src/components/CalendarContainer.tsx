import * as classNames from 'classnames';
import * as moment from 'moment';
import * as React from 'react';
import { IDatePicker } from '../common/@types';
import CalendarBody from './CalendarBody';
import CalendarHead from './CalendarHead';
import { Props as DayViewProps } from './DayView';
import TodayPanel from './TodayPanel';
import { ifExistCall } from '../utils/FunctionUtil';
import { DatePickerDefaults } from '../common/Constant';

interface CalendarContainerProps {
  /** Locale to use */
  locale?: string;
  /** Calendar Show or Hide */
  show?: boolean;
  /** PrevIcon Show or Hide */
  prevIcon?: boolean;
  /** NextIcon Show or Hide */
  nextIcon?: boolean;
  /** Event for Calendar day click */
  onChange?: (date: moment.Moment) => void;
  /** TodayPanel show or hide */
  showToday?: boolean;
}

interface PrivateProps {
  /** CalendarContainer base prop */
  current: moment.Moment;
  /** Default Date parameter in calendar, which is the parent component */
  base: moment.Moment;
  /** Number of months to show at once */
  showMonthCnt: number;
  /** Set Calendar initial Date  */
  setBase: (base: moment.Moment) => void;
}

export interface State {
  viewMode: IDatePicker.ViewMode;
}

export type InheritProps = DayViewProps & CalendarContainerProps;
export type Props = CalendarContainerProps & DayViewProps & PrivateProps;

class CalendarContainer extends React.Component<Props, State> {
  public static defaultProps = {
    current: moment(),
    show: true,
    showMonthCnt: 1,
    showToday: false,
    locale: DatePickerDefaults.locale,
  };

  public state = {
    viewMode: IDatePicker.ViewMode.DAY,
  };

  constructor(props: Props) {
    super(props);
    if (props.locale) {
      require(`moment/locale/${props.locale}`);
    }
  }

  public getHeaderTitle = () => {
    const { current } = this.props;
    const year = current.year();
    return {
      [IDatePicker.ViewMode.YEAR]: `${year - 4} - ${year + 5}`,
      [IDatePicker.ViewMode.MONTH]: `${year}`,
      [IDatePicker.ViewMode.DAY]: current.format('YYYY.MM'),
    }[this.state.viewMode];
  };

  public handleTitleClick = () => {
    const { viewMode } = this.state;
    const { showMonthCnt } = this.props;
    let changedMode: IDatePicker.ViewMode = viewMode;

    if (viewMode === IDatePicker.ViewMode.MONTH) {
      changedMode = IDatePicker.ViewMode.YEAR;
    } else if (viewMode === IDatePicker.ViewMode.DAY) {
      changedMode = IDatePicker.ViewMode.MONTH;
    }
    this.setState({
      viewMode: showMonthCnt > 1 ? IDatePicker.ViewMode.DAY : changedMode,
    });
  };

  public handleChange = (value: string) => {
    const { viewMode } = this.state;
    const { current, onChange, setBase, showMonthCnt, base } = this.props;
    if (!value.trim()) return;
    if (showMonthCnt > 1) {
      const date = current.date(parseInt(value, 10));
      ifExistCall(onChange, date);
      return;
    }

    if (viewMode === IDatePicker.ViewMode.YEAR) {
      setBase(base.clone().year(parseInt(value, 10)));
      this.setState({
        viewMode: IDatePicker.ViewMode.MONTH,
      });
    } else if (viewMode === IDatePicker.ViewMode.MONTH) {
      const month = moment()
        .month(value)
        .format('M');
      setBase(base.clone().month(parseInt(month, 10) - 1));
      this.setState({
        viewMode: IDatePicker.ViewMode.DAY,
      });
    } else {
      const date = current.date(parseInt(value, 10));
      ifExistCall(onChange, date);
    }
  };

  public handleBase = (method: string) => () => {
    const { base, setBase } = this.props;
    const { viewMode } = this.state;
    const clone = base.clone();
    if (viewMode === IDatePicker.ViewMode.YEAR) {
      setBase(clone[method](10, 'years'));
    } else if (viewMode === IDatePicker.ViewMode.MONTH) {
      setBase(clone[method](1, 'years'));
    } else {
      setBase(clone[method](1, 'months'));
    }
  };

  public handleToday = () => {
    const { setBase } = this.props;
    setBase(moment());
  };

  public render() {
    const {
      customDayClass,
      customDayText,
      disableDay,
      selected,
      showToday,
      startDay,
      endDay,
      prevIcon,
      nextIcon,
      show,
      onMouseOver,
      current,
      locale = DatePickerDefaults.locale,
    } = this.props;

    const calendarClass = classNames('calendar__container', {
      'calendar--show': show,
    });

    return (
      <div className={calendarClass}>
        <CalendarHead
          onPrev={this.handleBase('subtract')}
          onNext={this.handleBase('add')}
          prevIcon={prevIcon}
          nextIcon={nextIcon}
          onTitleClick={this.handleTitleClick}
          title={this.getHeaderTitle()}
        />
        {showToday && (
          <TodayPanel
            today={moment()
              .locale(locale)
              .format('LL')}
            onClick={this.handleToday}
            show={showToday}
          />
        )}
        <CalendarBody
          viewMode={this.state.viewMode}
          current={current}
          selected={selected}
          startDay={startDay}
          endDay={endDay}
          disableDay={disableDay}
          onClick={this.handleChange}
          onMouseOver={onMouseOver}
          customDayClass={customDayClass}
          customDayText={customDayText}
          locale={locale}
        />
      </div>
    );
  }
}

export default CalendarContainer;
