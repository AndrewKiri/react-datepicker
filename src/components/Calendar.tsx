import { range } from 'lodash';
import * as moment from 'moment';
import * as React from 'react';
import CalendarContainer, { InheritProps as ContainerProps } from './CalendarContainer';

export interface Props extends ContainerProps {
  base: moment.Moment;
  showMonthCnt: number;
}

export interface State {
  base: moment.Moment;
}

class Calendar extends React.Component<Props, State> {
  public static defaultProps = {
    base: moment(),
    showMonthCnt: 1,
    showToday: true,
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      base: props.base,
    };
  }

  public setBase = (base: moment.Moment) => {
    this.setState({ base });
  };

  public render() {
    const { showMonthCnt } = this.props;
    const { base } = this.state;

    return (
      <div className="calendar">
        <div className="calendar__list">
          {range(showMonthCnt).map(idx => (
            <div className="calendar__item" key={idx}>
              <CalendarContainer
                {...this.props}
                base={this.state.base}
                current={base.clone().add(idx, 'months')}
                prevIcon={idx === 0}
                nextIcon={idx === showMonthCnt! - 1}
                setBase={this.setBase}
              />
            </div>
          ))}
        </div>
      </div>
    );
  }
}

export default Calendar;
