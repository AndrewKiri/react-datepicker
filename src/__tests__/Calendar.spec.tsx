import { shallow } from 'enzyme';
import * as moment from 'moment';
import * as React from 'react';
import * as sinon from 'sinon';
import Calendar from '../Calendar';
import CalendarContainer from '../CalendarContainer';

const mockMoment = moment.unix(1543622400);

describe('<Calendar/>', () => {
  it('redners with no props', () => {
    const component = shallow(<Calendar base={mockMoment} />);

    expect(component).toMatchSnapshot();
  });

  it('props showMonthCnt correctly', () => {
    const component = shallow(<Calendar base={mockMoment} showMonthCnt={3} />);

    expect(component).toMatchSnapshot();
    expect(component.find('.calendar__item')).toHaveLength(3);
    // first calendar only previcon true
    expect(
      component
        .find(CalendarContainer)
        .first()
        .props().prevIcon
    ).toBeTruthy();
    // last calendar only nextIcon true
    expect(
      component
        .find(CalendarContainer)
        .last()
        .props().nextIcon
    ).toBeTruthy();
    // another calendar both hide
    expect(
      component
        .find(CalendarContainer)
        .at(1)
        .props().nextIcon
    ).toBeFalsy();
    expect(
      component
        .find(CalendarContainer)
        .at(1)
        .props().prevIcon
    ).toBeFalsy();
  });

  it('props top, left correctly', () => {
    const component = shallow(<Calendar base={mockMoment} top="100px" left="100px" />);

    expect(component).toMatchSnapshot();
    expect(component.props().style).toHaveProperty('top', '100px');
    expect(component.props().style).toHaveProperty('left', '100px');
  });
});
