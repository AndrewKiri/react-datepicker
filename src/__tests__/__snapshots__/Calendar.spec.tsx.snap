// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar/> props showMonthCnt correctly 1`] = `
<div
  className="calendar"
  style={
    Object {
      "left": undefined,
      "top": undefined,
    }
  }
>
  <div
    className="calendar__list"
  >
    <div
      className="calendar__item"
      key="0"
    >
      <CalendarContainer
        base={"2018-12-01T00:00:00.000Z"}
        current={"2018-12-01T00:00:00.000Z"}
        headerFormat="YYYY년 MM월"
        nextIcon={false}
        onNext={[Function]}
        onPrev={[Function]}
        prevIcon={true}
        show={true}
        showMonthCnt={3}
      />
    </div>
    <div
      className="calendar__item"
      key="1"
    >
      <CalendarContainer
        base={"2018-12-01T00:00:00.000Z"}
        current={"2019-01-01T00:00:00.000Z"}
        headerFormat="YYYY년 MM월"
        nextIcon={false}
        onNext={[Function]}
        onPrev={[Function]}
        prevIcon={false}
        show={true}
        showMonthCnt={3}
      />
    </div>
    <div
      className="calendar__item"
      key="2"
    >
      <CalendarContainer
        base={"2018-12-01T00:00:00.000Z"}
        current={"2019-02-01T00:00:00.000Z"}
        headerFormat="YYYY년 MM월"
        nextIcon={true}
        onNext={[Function]}
        onPrev={[Function]}
        prevIcon={false}
        show={true}
        showMonthCnt={3}
      />
    </div>
  </div>
</div>
`;

exports[`<Calendar/> props top, left correctly 1`] = `
<div
  className="calendar"
  style={
    Object {
      "left": "100px",
      "top": "100px",
    }
  }
>
  <div
    className="calendar__list"
  >
    <div
      className="calendar__item"
      key="0"
    >
      <CalendarContainer
        base={"2018-12-01T00:00:00.000Z"}
        current={"2018-12-01T00:00:00.000Z"}
        headerFormat="YYYY년 MM월"
        left="100px"
        nextIcon={true}
        onNext={[Function]}
        onPrev={[Function]}
        prevIcon={true}
        show={true}
        showMonthCnt={1}
        top="100px"
      />
    </div>
  </div>
</div>
`;

exports[`<Calendar/> redners with no props 1`] = `
<div
  className="calendar"
  style={
    Object {
      "left": undefined,
      "top": undefined,
    }
  }
>
  <div
    className="calendar__list"
  >
    <div
      className="calendar__item"
      key="0"
    >
      <CalendarContainer
        base={"2018-12-01T00:00:00.000Z"}
        current={"2018-12-01T00:00:00.000Z"}
        headerFormat="YYYY년 MM월"
        nextIcon={true}
        onNext={[Function]}
        onPrev={[Function]}
        prevIcon={true}
        show={true}
        showMonthCnt={1}
      />
    </div>
  </div>
</div>
`;
